main.go

```diff
@@ -21,54 +21,193 @@
+		"Directory whose descendants contain all the used Thrift files. "+
+			"The structure of the generated Go packages mirrors the paths to "+
+			"the Thrift files relative to this directory. By default, this is "+
+			"the deepest common ancestor of the Thrift files.")
+
+	var noRecurse bool
+	flag.BoolVar(&noRecurse, "no-recurse", false,
+		"Disable code generation for included Thrift files.")
 
-	output := flag.String("o", "", "Output file")
 	flag.Parse()
-	file := flag.Arg(0)
 
-	outDir, err := os.Getwd()
+	inputFile := flag.Arg(0)
+	if len(inputFile) == 0 {
+		flag.Usage()
+		os.Exit(1)
+	}
+	if _, err := os.Stat(inputFile); err != nil {
+		if os.IsNotExist(err) {
+			log.Fatalf("file %q does not exist: %v", inputFile, err)
+		}
+		log.Fatalf("error describing file %q: %v", inputFile, err)
+	}
+
+	if len(outputDir) == 0 {
+		outputDir = "."
+	}
+	outputDir, err := filepath.Abs(outputDir)
 	if err != nil {
-		log.Fatal(err)
+		log.Fatalf("could not resolve path %q: %v", outputDir, err)
 	}
 
-	if len(*output) > 0 {
-		var err error
-		if *output, err = filepath.Abs(*output); err != nil {
-			log.Fatal(err)
+	if packagePrefix == "" {
+		packagePrefix, err = determinePackagePrefix(outputDir)
+		if err != nil {
+			log.Fatalf("could not determine the package prefix: %v", err)
 		}
-		outDir = filepath.Dir(*output)
 	}
 
-	module, err := compile.Compile(file)
+	module, err := compile.Compile(inputFile)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	outFile := os.Stdout
-	if len(*output) > 0 {
-		outFile, err = os.Create(*output)
-		defer outFile.Close()
+	if thriftRoot == "" {
+		thriftRoot, err = findCommonAncestor(module)
 		if err != nil {
 			log.Fatal(err)
 		}
+	} else {
+		thriftRoot, err = filepath.Abs(thriftRoot)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if err := verifyAncestry(module, thriftRoot); err != nil {
+			log.Fatal(err)
+		}
 	}
 
 	opts := gen.Options{
-		PackageName: filepath.Base(outDir), // TODO: customize using flags
-		Output:      outFile,
+		OutputDir:     outputDir,
+		PackagePrefix: packagePrefix,
+		ThriftRoot:    thriftRoot,
+		NoRecurse:     noRecurse,
 	}
+
 	if err := gen.Generate(module, &opts); err != nil {
 		log.Fatal(err)
 	}
 }
+
+// verifyAncestry verifies that the Thrift file for the given module and the
+// Thrift files for all imported modules are contained within the directory
+// tree rooted at the given path.
+func verifyAncestry(m *compile.Module, root string) error {
+	return m.Walk(func(m *compile.Module) error {
+		path, err := filepath.Rel(root, m.ThriftPath)
+		if err != nil {
+			return fmt.Errorf(
+				"could not resolve path for %q: %v", m.ThriftPath, err)
+		}
+
+		if strings.HasPrefix(path, "..") {
+			return fmt.Errorf(
+				"%q is not contained in the %q directory tree",
+				m.ThriftPath, root)
+		}
+
+		return nil
+	})
+}
+
+// findCommonAncestor finds the deepest common ancestor for the Thrift file
+// for the given module and all the modules imported by it.
```

Comment:
the method doc is a little confusing "for the Thrift file for the given module". Maybe just "for the given module and all modules imported by it" is enough